File: web/Task2/TCircle.js
class TCircle {
    #radius;
    constructor(radius, name = "TCircle") {
        if (radius < 0) throw new Error("Negative radius bro");
        this.#radius = radius;
        this.name = name;
    }
    getRadius() { return this.#radius; }
    setRadius(r) {
        if (r < 0) throw new Error("Negative radius bro");
        this.#radius = r;
    }

    getArea() { return Math.PI * this.getRadius() * this.getRadius(); }
    getSectorArea(angle) {
        const norm = angle % 360;
        const factor = norm === 0 ? 1 : norm / 360;
        return this.getArea() * factor;
    }
    getCircleLength() { return 2 * Math.PI * this.getRadius(); }
    toString() { return `Circle "${this.name}"(r=${this.getRadius()})`; }

    add(other) { return new this.constructor(this.getRadius() + (other.getRadius() ?? other), this.name + "+" + other.name); }
    sub(other) { return new this.constructor(Math.abs(this.getRadius() - (other.getRadius() ?? other)), this.name + "-" + other.name); }
    mul(other) { return new this.constructor(this.getRadius() * (other.getRadius() ?? other), this.name + "*" + other.name); }
    div(other) { return new this.constructor(this.getRadius() / (other.getRadius() ?? other), this.name + "/" + other.name); }

    eq(other) { return this.getRadius() === other.getRadius(); }
    neq(other) { return this.getRadius() !== other.getRadius(); }
    gt(other) { return this.getRadius() > other.getRadius(); }
    lt(other) { return this.getRadius() < other.getRadius(); }
    ge(other) { return this.getRadius() >= other.getRadius(); }
    le(other) { return this.getRadius() <= other.getRadius(); }
}

File: c#/oop1-lab1/Task2/TCircle.cs
using System;
using System.Numerics;
using System.Collections.Generic;

namespace Lab.Task2;

class TCircle : ICloneable
{
    private double _r;
    public double radius
    {
        get => _r;
        set => _r = value < 0 ? throw new ArgumentException("Negative radius...") : value;
    }
    public TCircle(){}
    public TCircle(double r) => radius = r;

    public object Clone() => new TCircle(radius);
    public override bool Equals(object obj) => obj is TCircle other && this.radius == other.radius;
    public override int GetHashCode() => radius.GetHashCode();
    public override string ToString() => radius.ToString();

    public virtual double GetArea() => Math.PI * radius * radius;
    public virtual double GetSectorArea(double angle) => (Math.PI * radius * radius) * ((angle%360)/360.0);
    public virtual double GetCircleLength() => Math.PI * 2 * radius;

    public static bool operator ==(TCircle a, TCircle b) => a.radius == b.radius;
    public static bool operator !=(TCircle a, TCircle b) => a.radius != b.radius;
    public static bool operator >(TCircle a, TCircle b) => a.radius > b.radius;
    public static bool operator <(TCircle a, TCircle b) => a.radius < b.radius;
    public static bool operator <=(TCircle a, TCircle b) => a.radius <= b.radius;
    public static bool operator >=(TCircle a, TCircle b) => a.radius >= b.radius;

    public static TCircle operator +(TCircle a, TCircle b) => new TCircle(Math.Abs(a.radius + b.radius));
    public static TCircle operator +(double a, TCircle b) => new TCircle(Math.Abs(a + b.radius));
    public static TCircle operator +(TCircle a, double b) => new TCircle(Math.Abs(a.radius + b));

    public static TCircle operator -(TCircle a, TCircle b) => new TCircle(Math.Abs(a.radius - b.radius));
    public static TCircle operator -(double a, TCircle b) => new TCircle(Math.Abs(a - b.radius));
    public static TCircle operator -(TCircle a, double b) => new TCircle(Math.Abs(a.radius - b));

    public static TCircle operator *(TCircle a, TCircle b) => new TCircle(Math.Abs(a.radius * b.radius));
    public static TCircle operator *(double a, TCircle b) => new TCircle(Math.Abs(a * b.radius));
    public static TCircle operator *(TCircle a, double b) => new TCircle(Math.Abs(a.radius * b));

    public static TCircle operator /(TCircle a, TCircle b) => new TCircle(Math.Abs(a.radius / b.radius));
    public static TCircle operator /(double a, TCircle b) => new TCircle(Math.Abs(a / b.radius));
    public static TCircle operator /(TCircle a, double b) => new TCircle(Math.Abs(a.radius / b));
}
